### DISCUSSION TITLE
Prebuilt gcc 12.1.0 with Bitmanip

### DISCUSSION THREAD

**Discussion by vhdlnerd** [stnolting/neorv32#610] [General]
I've using the prebuilt gcc 12.1.0 (from [here](https://github.com/stnolting/riscv-gcc-prebuilt) - rv64imc-3.0.0 w/multilib).
In my makefile, I have this (before including the `common.mk` file):
```makefile
RISCV_PREFIX  = riscv64-unknown-elf-
MARCH        ?= rv32imc_zicsr
MABI         ?= ilp32
```
I have the neorv32 core built with the `m` and `c` options (and bitmanip). All my C code compiles/links fine -- no issues.

Now, I want to build with riscv bitmanip:
```makefile
RISCV_PREFIX  = riscv64-unknown-elf-
MARCH        ?= rv32imc_zicsr_zbb_zba_zbc_zbs
MABI         ?= ilp32
```
The C compiler runs fine but I get an error during linking. Gcc is trying to link with the 64-bit version of `libc.a` instead of the 32-bit `imc` version.

I'm I missing something (a gcc switch or environment variable setting) or is the prebuilt gcc 12.1.0 toolchain not built to support this?

Note: I can do a workaround by copying the 32-bit `imc` versions of `libc.a`, `libm.a` and `libgcc.a` to my local `lib` directory and force gcc to link against them. (So, I know its possible to build with the bitmanip features enabled.)  But, this "solution" is not ideal -- i.e., having toolchain libs in my local area.

Thanks,
Brian

**Comment by vhdlnerd** [comment]
I ended up building the GCC 13.x toolchain from source with just support for `rv32imc_zicsr_zbb_zba_zbc_zbs` and I'm using that now.

**Reply by stnolting** [reply]
Those multilib configuration are still exceeding my understanding of GCC's build system 😅

Just a note: you can also use the plain rv32i GCC toolchain and enable the different bit-manipulation extensions via the `misa` string. Thus, the compiler will compile your code emitting those additional instructions **but** the _pre-compiled_ compiler libraries (let's say "math.h") were - well - precompiled using the rv32i only.

> I ended up building the GCC 13.x toolchain from source with just support for rv32imc_zicsr_zbb_zba_zbc_zbs and I'm using that now.

:+1:

I think now that RISC-V GCC 13.1 is available I should update my prebuild toolchains...

**Reply by vhdlnerd** [reply]
You should wait a little on building RISC-V GCC 13.1, [riscv-gnu-toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain) is not supporting 13.1, yet but they will soon.  I did a "non-standard" 13.1 build because 12.2 had an issue build for `rv32imc_zicsr_zbb_zba_zbc_zbs` (see [issue](https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1236)).

> Those multilib configuration are still exceeding my understanding of GCC's build system

I don't know much about building toolchains, either.  I just followed directions found [here](https://github.com/riscv-collab/riscv-gnu-toolchain/issues/1249) to make my 13.1 build.

**Reply by stnolting** [reply]
> You should wait a little on building RISC-V GCC 13.1, [riscv-gnu-toolchain](https://github.com/riscv-collab/riscv-gnu-toolchain) is not supporting 13.1, yet but they will soon

Thanks for the hint!
